File: test/language/sql_virtual_tables/cpp/Fts3VirtualTableTest.cpp
#include <cstdio>
#include <string>
#include <vector>
#include <memory>

#include "gtest/gtest.h"

#include "sql_virtual_tables/fts3_virtual_table/Fts3TestDb.h"

#include "zserio/RebindAlloc.h"
#include "zserio/StringConvertUtil.h"
#include "zserio/SqliteFinalizer.h"

namespace sql_virtual_tables
{
namespace fts3_virtual_table
{

using allocator_type = Fts3TestDb::allocator_type;
using string_type = zserio::string<allocator_type>;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class Fts3VirtualTableTest : public ::testing::Test
{
public:
    Fts3VirtualTableTest()
    {
        std::remove(DB_FILE_NAME);

        m_database = new sql_virtual_tables::fts3_virtual_table::Fts3TestDb(DB_FILE_NAME);
        m_database->createSchema();
    }

    ~Fts3VirtualTableTest() override
    {
        delete m_database;
    }

    Fts3VirtualTableTest(const Fts3VirtualTableTest&) = delete;
    Fts3VirtualTableTest& operator=(const Fts3VirtualTableTest&) = delete;

    Fts3VirtualTableTest(Fts3VirtualTableTest&&) = delete;
    Fts3VirtualTableTest& operator=(Fts3VirtualTableTest&&) = delete;

protected:
    static void fillFts3VirtualTableRow(Fts3VirtualTable::Row& row, const string_type& title,
            const string_type& body)
    {
        row.setTitle(title);
        row.setBody(body);
    }

    static void fillFts3VirtualTableRows(vector_type<Fts3VirtualTable::Row>& rows)
    {
        rows.clear();
        for (int32_t id = 0; id < NUM_VIRTUAL_TABLE_ROWS; ++id)
        {
            const string_type title = "Title" + zserio::toString<allocator_type>(id);
            const string_type body = "Body" + zserio::toString<allocator_type>(id);
            Fts3VirtualTable::Row row;
            fillFts3VirtualTableRow(row, title, body);
            rows.push_back(row);
        }
    }

    static void checkFts3VirtualTableRow(const Fts3VirtualTable::Row& row1, const Fts3VirtualTable::Row& row2)
    {
        ASSERT_EQ(row1.getTitle(), row2.getTitle());
        ASSERT_EQ(row1.getBody(), row2.getBody());
    }

    static void checkFts3VirtualTableRows(const vector_type<Fts3VirtualTable::Row>& rows1,
            const vector_type<Fts3VirtualTable::Row>& rows2)
    {
        ASSERT_EQ(rows1.size(), rows2.size());
        for (size_t i = 0; i < rows1.size(); ++i)
            checkFts3VirtualTableRow(rows1[i], rows2[i]);
    }

    bool isTableInDb()
    {
        string_type checkTableName = "fts3VirtualTable";
        string_type sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + checkTableName +
                "'";
        std::unique_ptr<sqlite3_stmt, zserio::SqliteFinalizer> statement(
                m_database->connection().prepareStatement(sqlQuery));

        int result = sqlite3_step(statement.get());
        if (result == SQLITE_DONE || result != SQLITE_ROW)
            return false;

        const unsigned char* readTableName = sqlite3_column_text(statement.get(), 0);
        if (readTableName == nullptr || checkTableName != reinterpret_cast<const char*>(readTableName))
            return false;

        return true;
    }

    static const char* DB_FILE_NAME;
    static const int32_t NUM_VIRTUAL_TABLE_ROWS;

    sql_virtual_tables::fts3_virtual_table::Fts3TestDb* m_database;
};

const char* Fts3VirtualTableTest::DB_FILE_NAME = "language/sql_virtual_tables/fts3_virtual_table_test.sqlite";
const int32_t Fts3VirtualTableTest::NUM_VIRTUAL_TABLE_ROWS = 5;

TEST_F(Fts3VirtualTableTest, deleteTable)
{
    ASSERT_TRUE(isTableInDb());

    Fts3VirtualTable& testTable = m_database->getFts3VirtualTable();
    testTable.deleteTable();
    ASSERT_FALSE(isTableInDb());

    testTable.createTable();
    ASSERT_TRUE(isTableInDb());
}

TEST_F(Fts3VirtualTableTest, readWithoutCondition)
{
    Fts3VirtualTable& testTable = m_database->getFts3VirtualTable();

    vector_type<Fts3VirtualTable::Row> writtenRows;
    fillFts3VirtualTableRows(writtenRows);
    testTable.write(writtenRows);

    vector_type<Fts3VirtualTable::Row> readRows;
    auto reader = testTable.createReader();
    while (reader.hasNext())
        readRows.push_back(reader.next());
    checkFts3VirtualTableRows(writtenRows, readRows);
}

TEST_F(Fts3VirtualTableTest, readWithCondition)
{
    Fts3VirtualTable& testTable = m_database->getFts3VirtualTable();

    vector_type<Fts3VirtualTable::Row> writtenRows;
    fillFts3VirtualTableRows(writtenRows);
    testTable.write(writtenRows);

    const string_type condition = "body='Body1'";
    vector_type<Fts3VirtualTable::Row> readRows;
    auto reader = testTable.createReader(condition);
    while (reader.hasNext())
        readRows.push_back(reader.next());
    ASSERT_EQ(1, readRows.size());

    const size_t expectedRowNum = 1;
    checkFts3VirtualTableRow(writtenRows[expectedRowNum], readRows[0]);
}

TEST_F(Fts3VirtualTableTest, update)
{
    Fts3VirtualTable& testTable = m_database->getFts3VirtualTable();

    vector_type<Fts3VirtualTable::Row> writtenRows;
    fillFts3VirtualTableRows(writtenRows);
    testTable.write(writtenRows);

    const string_type updateTitle = "Title3";
    Fts3VirtualTable::Row updateRow;
    fillFts3VirtualTableRow(updateRow, updateTitle, "UpdatedName");
    const string_type updateCondition = "title='" + updateTitle + "'";
    testTable.update(updateRow, updateCondition);

    vector_type<Fts3VirtualTable::Row> readRows;
    auto reader = testTable.createReader(updateCondition);
    while (reader.hasNext())
        readRows.push_back(reader.next());
    ASSERT_EQ(1, readRows.size());

    checkFts3VirtualTableRow(updateRow, readRows[0]);
}

} // namespace fts3_virtual_table
} // namespace sql_virtual_tables

File: test/language/sql_virtual_tables/java/sql_virtual_tables/fts3_virtual_table/Fts3VirtualTableTest.java
package sql_virtual_tables.fts3_virtual_table;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import java.io.File;
import java.io.IOException;
import java.net.URISyntaxException;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import org.junit.After;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

import sql_virtual_tables.fts3_virtual_table.Fts3TestDb;
import test_utils.FileUtil;
import test_utils.JdbcUtil;

import zserio.runtime.ZserioError;
import zserio.runtime.SqlDatabase.Mode;

public class Fts3VirtualTableTest
{
    @BeforeClass
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @Before
    public void setUp() throws IOException, URISyntaxException, SQLException
    {
        FileUtil.deleteFileIfExists(file);
        database = new Fts3TestDb(file.toString());
        database.createSchema();
    }

    @After
    public void tearDown() throws SQLException
    {
        if (database != null)
        {
            database.close();
            database = null;
        }
    }

    @Test
    public void deleteTable() throws SQLException
    {
        assertTrue(isTableInDb());

        final Fts3VirtualTable testTable = database.getFts3VirtualTable();
        testTable.deleteTable();
        assertFalse(isTableInDb());

        testTable.createTable();
        assertTrue(isTableInDb());
    }

    @Test
    public void readWithoutCondition() throws SQLException, URISyntaxException, IOException, ZserioError
    {
        final Fts3VirtualTable testTable = database.getFts3VirtualTable();

        final List<Fts3VirtualTableRow> writtenRows = new ArrayList<Fts3VirtualTableRow>();
        fillFts3VirtualTableRows(writtenRows);
        testTable.write(writtenRows);

        final List<Fts3VirtualTableRow> readRows = testTable.read();
        checkFts3VirtualTableRows(writtenRows, readRows);
    }

    @Test
    public void readWithCondition() throws SQLException, URISyntaxException, IOException, ZserioError
    {
        final Fts3VirtualTable testTable = database.getFts3VirtualTable();

        final List<Fts3VirtualTableRow> writtenRows = new ArrayList<Fts3VirtualTableRow>();
        fillFts3VirtualTableRows(writtenRows);
        testTable.write(writtenRows);

        final String condition = "body='Body1'";
        final List<Fts3VirtualTableRow> readRows = testTable.read(condition);

        assertEquals(1, readRows.size());

        final int expectedRowNum = 1;
        final Fts3VirtualTableRow readRow = readRows.get(0);
        checkFts3VirtualTableRow(writtenRows.get(expectedRowNum), readRow);
    }

    @Test
    public void update() throws SQLException, URISyntaxException, IOException, ZserioError
    {
        final Fts3VirtualTable testTable = database.getFts3VirtualTable();

        final List<Fts3VirtualTableRow> writtenRows = new ArrayList<Fts3VirtualTableRow>();
        fillFts3VirtualTableRows(writtenRows);
        testTable.write(writtenRows);

        final String updateTitle = "Title3";
        final Fts3VirtualTableRow updateRow = createFts3VirtualTableRow(updateTitle, "UpdatedBody");
        final String updateCondition = "title='" + updateTitle + "'";
        testTable.update(updateRow, updateCondition);

        final List<Fts3VirtualTableRow> readRows = testTable.read(updateCondition);
        assertEquals(1, readRows.size());

        final Fts3VirtualTableRow readRow = readRows.get(0);
        checkFts3VirtualTableRow(updateRow, readRow);
    }

    private static void fillFts3VirtualTableRows(List<Fts3VirtualTableRow> rows)
    {
        for (int id = 0; id < NUM_VIRTUAL_TABLE_ROWS; ++id)
        {
            rows.add(createFts3VirtualTableRow("Title" + id, "Body" + id));
        }
    }

    private static Fts3VirtualTableRow createFts3VirtualTableRow(String title, String body)
    {
        final Fts3VirtualTableRow row = new Fts3VirtualTableRow();
        row.setTitle(title);
        row.setBody(body);

        return row;
    }

    private static void checkFts3VirtualTableRows(List<Fts3VirtualTableRow> rows1,
            List<Fts3VirtualTableRow> rows2)
    {
        assertEquals(rows1.size(), rows2.size());
        for (int i = 0; i < rows1.size(); ++i)
            checkFts3VirtualTableRow(rows1.get(i), rows2.get(i));
    }

    private static void checkFts3VirtualTableRow(Fts3VirtualTableRow row1, Fts3VirtualTableRow row2)
    {
        assertEquals(row1.getTitle(), row2.getTitle());
        assertEquals(row1.getBody(), row2.getBody());
    }

    private boolean isTableInDb() throws SQLException
    {
        // check if database does contain table
        final String sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + TABLE_NAME +
                "'";

        final PreparedStatement statement = database.prepareStatement(sqlQuery);
        try
        {
            final ResultSet resultSet = statement.executeQuery();
            if (!resultSet.next())
                return false;

            // read table name
            final String tableName = resultSet.getString(1);
            if (resultSet.wasNull() || !tableName.equals(TABLE_NAME))
                return false;
        }
        finally
        {
            statement.close();
        }

        return true;
    }

    private static final String TABLE_NAME = "fts3VirtualTable";

    private static final int    NUM_VIRTUAL_TABLE_ROWS = 5;

    private static final String FILE_NAME = "fts3_virtual_table_test.sqlite";

    private final File file = new File(FILE_NAME);
    private Fts3TestDb database = null;
}
